src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedEntryPointSymbol.cs

```diff
@@ -329,6 +334,160 @@ private static BoundCall CreateParameterlessCall(CSharpSyntaxNode syntax, BoundE
+
+                _userMain = userMain;
+                _paramKind = userMain.ParameterCount == 0 ? ParamKind.NoArgs : ParamKind.StringArrayArgs;
+                _returnKind = GetReturnKind(compilation, userMain.ReturnType);
+
+                switch (_paramKind)
+                {
+                    case ParamKind.StringArrayArgs:
+                        var stringType = compilation.GetSpecialType(SpecialType.System_String);
+                        var stringArrayType = ArrayTypeSymbol.CreateCSharpArray(compilation.Assembly, stringType);
+                        _parameters = ImmutableArray.Create(
+                            SynthesizedParameterSymbol.Create(this, stringArrayType, 0, RefKind.None, userMain.Parameters[0].Name));
+                        break;
+                    default:
+                        _parameters = ImmutableArray<ParameterSymbol>.Empty;
+                        break;
+                }
+
+                var userMainLocation = userMain.DeclaringSyntaxReferences.SingleOrDefault()?.GetLocation() ?? NoLocation.Singleton;
+
+                _getAwaiterMethod = GetRequiredMethod(_userMain.ReturnType, WellKnownMemberNames.GetAwaiter, diagnosticBag, userMainLocation);
+                if ((object)_getAwaiterMethod != null)
+                {
+                    _getResultMethod = GetRequiredMethod(_getAwaiterMethod.ReturnType, WellKnownMemberNames.GetResult, diagnosticBag);
+                }
+            }
+
+            public override string Name => MainName;
+
+            public override ImmutableArray<ParameterSymbol> Parameters => _parameters;
+
+            internal override BoundBlock CreateBody()
+            {
+                var syntax = this.GetSyntax();
+
+                var arguments = Parameters.SelectAsArray(p => new BoundParameter(syntax, p, p.Type) as BoundExpression);
+
+                // Main(args) or Main()
+                BoundCall userMainInvocation = new BoundCall(
+                        syntax: this.GetSyntax(),
+                        receiverOpt: null,
+                        method: _userMain,
+                        arguments: arguments,
+                        argumentNamesOpt: default(ImmutableArray<string>),
+                        argumentRefKindsOpt: default(ImmutableArray<RefKind>),
+                        isDelegateCall: false,
+                        expanded: false,
+                        invokedAsExtensionMethod: false,
+                        argsToParamsOpt: default(ImmutableArray<int>),
+                        resultKind: LookupResultKind.Viable,
+                        type: _userMain.ReturnType)
+                { WasCompilerGenerated = true };
+
+
+                // GetAwaiter().GetResult()
+                BoundCall getAwaiterGetResult =
+                    CreateParameterlessCall(
+                            syntax: syntax,
+                            method: _getResultMethod,
```

Comment:
Can CreateBody be called if the constructor for AsyncForwardEntryPoint added an error to the diagnostic bag?  If so this would end up passing a `null` value for `_getResultMethod`.